src/python/Services/ReqMgr/ReqMgrReader.py

```diff
@@ -6,53 +6,383 @@
-        except Exception as e:
+
+        except Exception as error:
             msg = "Error initializing ReqMgrReader\n"
-            msg += "%s\n" % format(e)
+            msg += f"{error}\n"
             raise Exception(msg)
 
-    def getWorkflowByCampaign(self, campaign, details=False):
+    def _runWithRetries(
+        self,
+        fcn: Callable,
+        fcn_pargs: list,
+        fcn_args: dict = {},
+        default: Optional[Any] = None,
+        tries: int = 10,
+        wait: int = 5,
+    ) -> Any:
+        # TODO: is there a better place to put this function?
+        """
+        The function to run a given function with retries
+        :param fcn: function
+        :param fcn_pargs: function arguments
+        :param fnc_args: optional function arguments
+        :default: default value to return in case all the tries have failed, raise Exception o/w
+        :tries: number of tries
+        :wait: wait time between tries
+        :return: function output, deafault value o/w
+        """
+        for i in range(tries):
+            try:
+                return fcn(*fcn_pargs, **fcn_args)
+
+            except Exception as error:
+                self.logger.error(
+                    f"Failed to get to run function {fcn.__name__} with arguments {fcn_pargs} and {fcn_args} on try #{i+1} of {tries}"
+                )
+                self.logger.error(str(error))
+                time.sleep(wait)
+
+        if default:
+            return default
+        raise Exception("NoDefaultValue")
+
+    def getWorkflowSchema(
+        self,
+        wf: str,
+        drop_null: bool = False,
+        make_copy: bool = False,
+        tries: int = 2,
+        try_cache: bool = False,
+        wait: int = 0,
+    ) -> dict:
+        """
+        The function to get the schema for a given workflow
+        :param wf: workflow name
+        :param drop_null: if True, drop key with null values from the schema
+        :param make_copy: if True, return a copy of the schema
+        :param tries: number of tries for gettting a request response
+        :return: a dict
+        """
+        try:
+            return self._runWithRetries(
+                self._getWorkflowSchema,
+                [wf],
+                {"drop_null": drop_null, "make_copy": make_copy},
+                tries=tries,
+                wait=wait,
+            )
+
+        except Exception as error:
+            self.logger.error(f"Failed to get workload from reqmgr for workflow {wf}")
+            self.logger.error(str(error))
+            if try_cache:
+                self.logger.info("Will try to get workload cache")
+                return self._getWorkflowSchema(
+                    wf, drop_null=drop_null, make_copy=make_copy, from_cache=True
+                )
+
+    def _getWorkflowSchema(
+        self,
+        wf: str,
+        drop_null: bool = False,
+        make_copy: bool = False,
+        from_cache: bool = False,
```

Comment:
I would also remove this `from_cache` parameter. Let's do not apply caching for now. We can handle it later.